packages/block-editor/src/utils/use-can-block-toolbar-be-focused.js

```diff
@@ -2,47 +2,76 @@
+	isUnmodifiedDefaultBlock,
+} from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import { store as blockEditorStore } from '../store';
 import { unlock } from '../lock-unlock';
+import { useHasAnyBlockControls } from '../components/block-controls/use-has-block-controls';
 
 /**
  * Returns true if the block toolbar should be able to receive focus.
  *
  * @return {boolean} Whether the block toolbar should be able to receive focus
  */
 export function useCanBlockToolbarBeFocused() {
-	return useSelect( ( select ) => {
-		const {
-			__unstableGetEditorMode,
-			getBlock,
-			getSettings,
-			getSelectedBlockClientId,
-			getFirstMultiSelectedBlockClientId,
-		} = unlock( select( blockEditorStore ) );
+	const hasAnyBlockControls = useHasAnyBlockControls();
 
-		const selectedBlockId =
-			getFirstMultiSelectedBlockClientId() || getSelectedBlockClientId();
-		const isEmptyDefaultBlock = isUnmodifiedDefaultBlock(
-			getBlock( selectedBlockId ) || {}
-		);
+	const { isToolbarEnabled, isDefaultEditingMode, maybeShowBlockToolbar } =
+		useSelect( ( select ) => {
+			const {
+				__unstableGetEditorMode,
+				getBlock,
+				getBlockName,
+				getBlockEditingMode,
+				getSettings,
+				getSelectedBlockClientId,
+				getFirstMultiSelectedBlockClientId,
+			} = unlock( select( blockEditorStore ) );
 
-		// Fixed Toolbar can be focused when:
-		// - a block is selected
-		// - fixed toolbar is on
-		// Block Toolbar Popover can be focused when:
-		// - a block is selected
-		// - we are in edit mode
-		// - it is not an empty default block
-		return (
-			!! selectedBlockId &&
-			( getSettings().hasFixedToolbar ||
-				( __unstableGetEditorMode() === 'edit' &&
-					! isEmptyDefaultBlock ) )
-		);
-	}, [] );
+			const selectedBlockId =
+				getFirstMultiSelectedBlockClientId() ||
+				getSelectedBlockClientId();
+			const isEmptyDefaultBlock = isUnmodifiedDefaultBlock(
+				getBlock( selectedBlockId ) || {}
+			);
+			const blockType =
+				selectedBlockId &&
+				getBlockType( getBlockName( selectedBlockId ) );
+
+			// Fixed Toolbar can be focused when:
+			// - a block is selected
+			// - the block has tools
+			// - fixed toolbar is on
+			// Block Toolbar Popover can be focused when:
+			// - a block is selected
+			// - the block has tools
+			// - we are in edit mode
+			// - it is not an empty default block
+			return {
+				isToolbarEnabled:
+					blockType &&
+					hasBlockSupport( blockType, '__experimentalToolbar', true ),
+				isDefaultEditingMode:
+					getBlockEditingMode( selectedBlockId ) === 'default',
+				maybeShowBlockToolbar:
+					getSettings().hasFixedToolbar ||
+					( __unstableGetEditorMode() === 'edit' &&
+						! isEmptyDefaultBlock ),
+			};
+		}, [] );
+
+	// The same check used in <BlockToolbar /> to see if it should return null.
+	// Should we combine these into their own hook so they stay consistent?
+	const noBlockToolbar =
+		! isToolbarEnabled ||
+		( ! isDefaultEditingMode && ! hasAnyBlockControls );
```

Comment:
I think having the "should we show the block toolbar" check within `<BlockToolbar />` and this check should be combined into a new "showBlockToolbar" hook that can get used here.